tools/ramdump/ramdump_tool.c

```diff
@@ -80,69 +91,201 @@ static int b_read(int fd, uint8_t *buf, int size)
+	/* Receive memory address, size & heap index for memory regions from TARGET */
+	for (i = 2; i <= number_of_regions + 1; i++) {
+		ret = b_read(dev_fd, (uint8_t *)&mem_address, 4);
+		if (ret != 4) {
+			printf("Receiving address failed, ret = %d\n", ret);
+			return -1;
+		}
+		mem_info[i].rd_regionx_start = mem_address;
+
+		ret = b_read(dev_fd, (uint8_t *)&mem_size, 4);
+		if (ret != 4) {
+			printf("Receiving size failed, ret = %d\n", ret);
+			return -1;
+		}
+		mem_info[i].rd_regionx_size = mem_size;
+
+		ret = read(dev_fd, &c, 1);
+		if (ret != 1) {
+			printf("Receiving size failed, ret = %d\n", ret);
+			return -1;
+		}
+		mem_info[i].rd_regionx_idx = c;
+	}
+
+	return 0;
+}
 
-	snprintf(bin_file, BINFILE_NAME_SIZE, "ramdump_0x%08x_0x%08x.bin", ramdump_address, (ramdump_address + ramdump_size));
+static int send_region_info(int dev_fd, char *host_region)
+{
+	int ret;
 
-	bin_fp = fopen(bin_file, "w");
-	if (bin_fp == NULL) {
-		printf("%s create failed\n", bin_file);
+	/* Send region info to TARGET */
+	ret = write(dev_fd, host_region, 1);
+	if (ret != 1) {
+		printf("Sending region info failed, ret = %d\n", ret);
 		return -1;
 	}
 
-	printf("[>");
-	fflush(stdout);
+	return 0;
+}
 
-	while (ramdump_size) {
-		ret = read(dev_fd, &buf, 1);
-		if (ret != 1) {
-			printf("Receiving ramdump %dTH byte failed, ret = %d\n", count, ret);
-			fclose(bin_fp);
-			return -1;
+static int ramdump_recv(int dev_fd)
+{
+	char buf;
+	int i;
+	int ret;
+	int index;
+	int count = 0;
+	int regions_to_dump = 0;
+	uint32_t ramdump_size;
+	char ramdump_region[2] = { '\0' };
+	char bin_file[BINFILE_NAME_SIZE] = { '\0' };
+	FILE *bin_fp;
+
+	/* Display memory region options for user to dump */
+	printf("\n=========================================================================\n");
+	printf("Ramdump Region Options:\n");
+	printf("1. ALL");
+	if (number_of_regions == 1) {
+		printf("\t( Address: %08x, Size: %d )\n", mem_info[2].rd_regionx_start, mem_info[2].rd_regionx_size);
+	} else {
+		for (index = 2; index <= number_of_regions + 1; index++) {
+			printf("\n%d. Region %d:\t( Address: 0x%08x, Size: %d )\t [Heap index = %d]", index, index - 2, mem_info[index].rd_regionx_start, mem_info[index].rd_regionx_size, mem_info[index].rd_regionx_idx);
 		}
+	}
+	printf("\n=========================================================================\n");
+	printf("Please enter desired ramdump option as below:\n \t1 for ALL\n");
+	if (number_of_regions > 1) {
+	printf(" \t2 for Region 0\n \t25 for Region 0 & 3 ...\n");
+	}
 
-		ret = fwrite(&buf, 1, 1, bin_fp);
-		if (ret != 1) {
-			printf("Writing ramdump %dTH byte failed, ret = %d\n", count, ret);
-			fclose(bin_fp);
-			return -1;
+	/* Take user's input for desired ramdump region */
+	printf("Please enter your input: ");
+	scanf("%d", &index);
+
+scan_input:
+	/* Check if user's input is valid */
+	while (index < 0 || index == EOF || (number_of_regions == 1 && index > 1)) {
```

Comment:
If the number of regions is 1, can't "2" be passed as a parameter?

Reply:
no, because that's not valid, so there will be a correct input prompt until user enters correct options